### PR TITLE
[ColossalChat] Colossalchat upgrade

### PR DESCRIPTION
**Pull request by YeAnbang**
## 📌 Checklist before creating the PR

- [ ] I have created an issue for this PR for traceability
- [ ] The title follows the standard format: `[doc/gemini/tensor/...]: A concise description`
- [ ] I have added relevant tags if possible for us to better distinguish different PRs
- [x] I have installed pre-commit: `pip install pre-commit && pre-commit install`

## 🚨 Issue number

> Link this PR to your issue with words like fixed to automatically close the linked issue upon merge
>
> e.g. `fixed #1234`, `closed #1234`, `resolved #1234`

## 📝 What does this PR do?

> Summarize your work here.
> if you have any plots/diagrams/screenshots/tables, please attach them here.

## 💥 Checklist before requesting a review

- [ ] I have linked my PR to an issue ([instruction](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue))
- [ ] My issue clearly describes the problem/feature/proposal, with diagrams/charts/table/code if possible
- [x] I have performed a self-review of my code
- [x] I have added thorough tests.
- [x] I have added docstrings for all the functions/methods I implemented

## ⭐️ Do you enjoy contributing to Colossal-AI?

- [x] 🌝 Yes, I do.
- [ ] 🌚 No, I don't.

Tell us more if you don't enjoy contributing to Colossal-AI.

### CODE REVIEW

**Review by TongLi3701 — CHANGES_REQUESTED** [review]
Thanks Anbang.

Some comments. Please also update corresponding README file and also add a table for supported models.

File: applications/ColossalChat/coati/dataset/tokenization_utils.py [hunk 1]
```diff
@@ -123,6 +135,20 @@ def supervised_tokenize_sft(
...
+            if labels[i] == ignore_index:
+                if start != end:
+                    label_decode.append(tokenizer.decode(labels[start + 1 : i], skip_special_tokens=False))
+                start = i
+                end = i
+            else:
+                end = i
+                if i == len(labels) - 1:
+                    label_decode.append(tokenizer.decode(labels[start + 1 :], skip_special_tokens=False))
+
```
[the remark this answers is not in the document]
Comment (YeAnbang) [reply applications/ColossalChat/coati/dataset/tokenization_utils.py:151]: removed

File: applications/ColossalChat/coati/dataset/loader.py [hunk 2]
```diff
@@ -345,22 +348,30 @@ def setup_distributed_dataloader(
...
     _kwargs = kwargs.copy()
     process_group = process_group or _get_default_group()
+    # world_size = tp_size * pp_size
+    assert (
+        process_group.size() % tp_size == 0
+    ), f"process_group.size()={process_group.size()} must be divisible by tp_size={tp_size}"
     sampler = StatefulDistributedSampler(
         dataset=dataset,
-        num_replicas=process_group.size() if not use_tp else 1,
+        num_replicas=int(process_group.size() / tp_size),
```
[the remark this answers is not in the document]
Comment (YeAnbang) [reply applications/ColossalChat/coati/dataset/loader.py:367]: changed to plugin.prepare_dataloader

File: applications/ColossalChat/config/conversation_template/deepseek-ai_DeepSeek-V2-Lite.json [hunk 3]
```diff
@@ -0,0 +1,8 @@
+{
+    "chat_template": "{% if not add_generation_prompt is defined %}{% set add_generation_prompt = false %}{% endif %}{{ bos_token }}{% for message in messages %}{% if message['role'] == 'user' %}{{ 'User: ' + message['content'] + '\n\n' }}{% elif message['role'] == 'assistant' %}{{ 'Assistant: ' + message['content'] + eos_token }}{% elif message['role'] == 'system' %}{{ message['content'] + '\n\n' }}{% endif %}{% endfor %}{% if add_generation_prompt %}{{ 'Assistant:' }}{% endif %}",
+    "system_message": "A chat between a curious human and an artificial intelligence assistant. The assistant gives helpful, detailed, and polite answers to the human's questions.\n\n",
+    "stop_ids": [
+        100001
+    ],
+    "end_of_assistant": "<｜end▁of▁sentence｜>"
+}
```
[the remark this answers is not in the document]
Comment (YeAnbang) [reply applications/ColossalChat/config/conversation_template/deepseek-ai_DeepSeek-V2-Lite.json:8]: will test it again

[the remark this answers is not in the document]
Comment (YeAnbang) [reply applications/ColossalChat/config/conversation_template/deepseek-ai_DeepSeek-V2-Lite.json:8]: {"messages": [{"from": "human", "content": "what are some pranks with a pen I can do?"}, {"from": "assistant", "content": "Are you looking for practical joke ideas?"}, {"from": "human", "content": "Yes, exactly!"}, {"from": "assistant", "content": "You could try attaching a small buzzer to a pen. When someone tries to use it, they'll be surprised by the sound!"}, {"from": "human", "content": "That's a funny one! Any other ideas?"}, {"from": "assistant", "content": "Another idea is the classic 'ink explosion' prank. Replace the ink of a pen with disappearing ink, and watch the confusion when the ink vanishes from the paper."}], "input_ids": [100000, 32, 12465, 1439, 245, 13076, 3807, 285, 274, 18050, 15141, 20308, 13, 429, 20308, 4380, 9394, 11, 9333, 11, 285, 30513, 9789, 276, 254, 3807, 6, 82, 4313, 13, 185, 185, 185, 185, 5726, 25, 856, 418, 742, 793, 3086, 366, 245, 5578, 304, 481, 536, 30, 185, 185, 77398, 25, 6475, 340, 2709, 327, 9743, 22539, 5121, 30, 100001, 5726, 25, 7587, 11, 5096, 0, 185, 185, 77398, 25, 1257, 1027, 1682, 51820, 245, 1760, 94613, 276, 245, 5578, 13, 2473, 3899, 18388, 276, 938, 359, 11, 657, 6, 647, 330, 12052, 457, 254, 3066, 0, 100001, 5726, 25, 2608, 6, 82, 245, 14700, 634, 0, 6295, 750, 5121, 30, 185, 185, 77398, 25, 11727, 3257, 317, 254, 11092, 655, 796, 30801, 6, 793, 1440, 13, 43470, 254, 21803, 280, 245, 5578, 366, 61484, 21803, 11, 285, 4348, 254, 16305, 754, 254, 21803, 31627, 473, 254, 3322, 13, 100001], "labels": [-100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, 6475, 340, 2709, 327, 9743, 22539, 5121, 30, 100001, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, 1257, 1027, 1682, 51820, 245, 1760, 94613, 276, 245, 5578, 13, 2473, 3899, 18388, 276, 938, 359, 11, 657, 6, 647, 330, 12052, 457, 254, 3066, 0, 100001, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, -100, 11727, 3257, 317, 254, 11092, 655, 796, 30801, 6, 793, 1440, 13, 43470, 254, 21803, 280, 245, 5578, 366, 61484, 21803, 11, 285, 4348, 254, 16305, 754, 254, 21803, 31627, 473, 254, 3322, 13, 100001], "inputs_decode": "<｜begin▁of▁sentence｜>A chat between a curious human and an artificial intelligence assistant. The assistant gives helpful, detailed, and polite answers to the human's questions.\n\n\n\nUser: what are some pranks with a pen I can do?\n\nAssistant: Are you looking for practical joke ideas?<｜end▁of▁sentence｜>User: Yes, exactly!\n\nAssistant: You could try attaching a small buzzer to a pen. When someone tries to use it, they'll be surprised by the sound!<｜end▁of▁sentence｜>User: That's a funny one! Any other ideas?\n\nAssistant: Another idea is the classic 'ink explosion' prank. Replace the ink of a pen with disappearing ink, and watch the confusion when the ink vanishes from the paper.<｜end▁of▁sentence｜>", "labels_decode": [" Are you looking for practical joke ideas?<｜end▁of▁sentence｜>", " You could try attaching a small buzzer to a pen. When someone tries to use it, they'll be surprised by the sound!<｜end▁of▁sentence｜>", " Another idea is the classic 'ink explosion' prank. Replace the ink of a pen with disappearing ink, and watch the confusion when the ink vanishes from the paper.<｜end▁of▁sentence｜>"], "seq_length": 151, "seq_category": "None"}

File: applications/ColossalChat/coati/models/critic.py [hunk 4]
```diff
@@ -32,3 +32,9 @@ def forward(self, input_ids: torch.LongTensor, attention_mask: Optional[torch.Te
         )
         values = self.value_head(sequence_hidden_states).squeeze(-1)  # ensure shape is (B, sequence length)
         return values
+
+    def get_input_embeddings(self):
+        return self.model.get_input_embeddings()
+
+    def get_output_embeddings(self):
+        return self.model.get_output_embeddings()
```
[the remark this answers is not in the document]
Comment (YeAnbang) [reply applications/ColossalChat/coati/models/critic.py:40]: no this is a class method, in colossalai 4.37.0, the boosted model specifically ask for these two methods, will probably remove those in future version when we also migrate to higher transformers version

File: applications/ColossalChat/examples/training_scripts/train_sft.py [hunk 5]
```diff
@@ -35,6 +42,43 @@ def train(args):
...
+                args.zero_stage = 2
+                if args.plugin == "zero2_cpu":
+                    args.zero_cpu_offload = True
+                else:
+                    args.zero_cpu_offload = False
+                logger.info(
+                    f"Model is compatible with hybrid parallel plugin, set plugin to {args.plugin} with zero_stage {args.zero_stage} and zero_cpu_offload {args.zero_cpu_offload}"
+                )
+    except NotImplementedError:
+        logger.warning(f"Unable to find a policy for the model, use {args.plugin} plugin instead")
```
Comment (TongLi3701) [inline applications/ColossalChat/examples/training_scripts/train_sft.py:70]: What is the aim of this policy check here?

File: applications/ColossalChat/examples/training_scripts/train_ppo.py [hunk 6]
```diff
@@ -261,27 +208,71 @@ def train(args):
...
+            sequence_parallelism_mode=args.sp_mode,
+            zero_stage=args.zero_stage,
+            enable_flash_attention=args.use_flash_attn,
+            enable_sequence_parallelism=args.enable_sequence_parallelism,
+            cpu_offload=True if args.zero_stage >= 1 and args.zero_cpu_offload else False,
             parallel_output=False,
+            max_norm=args.grad_clip,
             precision=args.mixed_precision,
-            custom_policy=booster_policy,
+            custom_policy=get_autopolicy(reward_model.model),
```
Comment (TongLi3701) [inline applications/ColossalChat/examples/training_scripts/train_ppo.py:239]: why do we need auto policy for reward and critic model here for 3d?

**Review by TongLi3701 — APPROVED** [review]
Thanks Anbang, overall looks fine.

I left some questions, please address them and merge.

File: applications/ColossalChat/examples/training_scripts/train_sft.py [hunk 7]
```diff
@@ -16,11 +18,16 @@
...
+from colossalai.logging import get_dist_logger
 from colossalai.nn.lr_scheduler import CosineAnnealingWarmupLR
 from colossalai.nn.optimizer import HybridAdam
 
+logger = get_dist_logger()
+
 
 def train(args):
+    print(colossalai.__version__, inspect.getfile(colossalai))
+    print(sys.executable)
```
[the remark this answers is not in the document]
Comment (YeAnbang) [reply applications/ColossalChat/examples/training_scripts/train_sft.py:30]: will remove it, it's for debugging

File: applications/ColossalChat/examples/training_scripts/train_ppo.py (hunk 6 above)
Comment (YeAnbang) [reply applications/ColossalChat/examples/training_scripts/train_ppo.py:239]: In the past, only 4 models are supported by colossal (e.g., bloom, gpt2...) and user needs to manually set the model type. Now, it is change to automatically select a policy (if supported)
